### ISSUE TITLE
train lora support

### ISSUE THREAD

**Issue by Yidhar** [OpenSenseNova/SenseNova-U1#114]
官方仓库有计划发布finetune / lora train脚本吗？
我在实现该模型的lora 训练的时候遇到了许多：如肢体崩坏，像素格纹的问题，想请教一下。

问题：
1. 对于下游风格 LoRA fine-tune，建议从哪个开始？ 从 final 版本继续训 还是从 SFT 版本开始训               
2. 可训练模块（trainable surface）
      我们做了系统的 ablation（4 个变体）后，目前的最优配方是：                                                                                              
                                                             
      Trainable:                                                                                                                                             
      - _mot_gen attention 4 个投影（q/k/v/o）× 42 layer = 168 个 LoRA 模块（rank=64, alpha=64）
      - _mot_gen MLP 3 个投影（gate/up/down）× 42 layer = 126 个 LoRA 模块（同 rank/alpha）                                                                  
      - fm_modules.timestep_embedder（全参数，~17.8M）                                     
      - fm_modules.noise_scale_embedder（全参数，~17.8M）                                                                                                    
                                                                                                                                                             
      合计 ~239M trainable params。              
                                                                                                                                                                                                                                                                        
      Frozen:                                                
      - vision_model（understand-side vision encoder）                                                                                                       
      - language_model.model.embed_tokens / lm_head                                                                                                          
      - ordinary path（非 _mot_gen 的 attn / mlp / norm）— 文本理解侧
      - language_model.model.norm / model.norm_mot_gen                                                                                                       
      - fm_modules.fm_head（输出 RGB-patch 的投影）                                                                                                          
      - fm_modules.vision_model_mot_gen（gen-side input patch encoder）        
       问题：                                                                                                                                                 
      - (a) RL 阶段官方训了哪些模块？我们目前是反向工程，希望得到"应该训哪些 / 不该训哪些"的指导。
      - (b) vision_model_mot_gen 和 fm_head 在 LoRA 风格 fine-tune 中是否可训？有无已知的风险（破坏 generation prior / quant 不兼容）？                      
      - (c)  训练过程中的注意事项
3. 训练超参
     - timestep / noise_scale 调度沿用官方 inference（timestep_shift=3.0, cfg_scale=4.0, cfg_norm="none"）                                                  
      - x0-MSE flow-matching loss（x0 ∈ [-1,1]，per official extract_feature normalize）                                                                     
      - noise_scale per-image: clamp(sqrt(N/64), max=8) × 1.0 base                                                                                           
      - prompt template: 走 model._build_t2i_query (chat template + system message + <think></think><img>)                                                   
      - snap-to-bucket: 每张训练图 resize 到最近的 11 个官方 bucket 之一（per examples/README.md "Supported resolution buckets"）   
      -  1. x0 normalize 空间：(0.5, 0.5, 0.5)/(0.5, 0.5, 0.5) 把 [0,1] 映射到 [-1, 1]                                                                          
      -  Per-image noise_scale：clamp(sqrt(image_token_num / 64), max=8)                                                                                     
      -  Prompt prefix：必须走 model._build_t2i_query(prompt, system_message=SYSTEM_MESSAGE_FOR_GEN, ...) 的 chat template
      -  Train shape 是否必须 ≡ infer shape 的 11 个官方 bucket（任意 round-to-32 的 ratio 是 OOD）                                                              
      -   ts_embedder / ns_embedder 可能是 scene-atmosphere conditioning（contrast / palette / rim light）的核心？
      
      - 尤其是loss计算这一段 我不确认是否正确
      
      -  我们使用 x0-prediction MSE（rectified-flow 风格）：           
                                                                                                                                      
      -    Per training step (single sample, batch=1):         
                                                                                                     
      -   t = uniform(t_eps, 1.0)                        # t_eps = 1e-3 typical                                                                                  
      -   eps ~ N(0, I)                                                                                                                                          
      -   z_t = t * x0_patch + (1 - t) * eps              # rectified-flow / OT linear interp   (t=0 噪声, t=1 干净)                                                                                 
      -   x_pred = fm_head(model_forward(z_t, prompt, ...))  # gen tower 直接预测 x0                                                                             
      -   loss = F.mse_loss(x_pred.float(), x0_patch.float())   # raw MSE, 无 SNR weighting          
                                                                                                                                                                                                                                
          其中：                                                                                                                                                 
          - x0_patch 是把 RGB 图按 32×32 patchify 后展平的 (N, 32*32*3=3072) 张量，已经 normalize 到 [-1, 1]（用 mean=0.5, std=0.5）                             
          - eps 是与 x0 同形 standard normal noise                                                                                                               
          - timestep t per step 单一标量（per-sample uniform 采样）
          - 没有引入 timestep-dependent weighting（如 logit-normal sampling, SNR-weighting, MinSNR-γ 等）                                                        
          - 没有同时算 v-prediction loss（虽然推理时 sampling 用的是 v）                                                                                         
          - gradient accumulation = 1（每 step 立即更新）                                                                                                        
                                                                                                                                                                 
          问题：                                                                                                                                                 
          - (a) 官方 RL 阶段是 x0-MSE 还是 v-MSE 还是 mixed？我们走 x0-MSE 是因为 _t2i_predict_v 内部就是用 x_pred 重建 velocity，但不知道训练阶段是否 v-MSE     
          更稳定。                                                                                                                                               
          - (b) Timestep 采样分布：我们用 uniform (t_eps, 1]，是否官方推荐 logit-normal / shifted / SNR-aware？尤其在做 finetune 阶段，是否需要偏向某一段 t      
          区间？                                                                                                                                                 
          - (c) Loss weighting：是否需要 1 / (1-t)^2 之类与 timestep 相关的 weight，或者 MinSNR-γ 处理？x0-MSE 直接 MSE 在小 t 区间（接近 x0）会自然 smaller loss
           magnitude。                                                                                                                                           
          - (d) 我们在 patch-level (N, 3072) 上算 mean MSE，是否官方在 pixel level 算？尺度差别会影响 effective LR。                                             
                                                                                                                    
          如果有官方训练时使用的 loss 公式（含 normalize 系数 / timestep weight）会非常有帮助。

**Comment by Paranioar** [comment]
感谢这么详细的反馈和实验分享🙏🙏

关于大家比较关心的训练细节，我们最近刚刚发布了[技术报告](https://github.com/OpenSenseNova/SenseNova-U1/blob/main/docs/pdf/SenseNOVA_U1.pdf)，里面对部分训练设计做一些说明，希望能解答部分疑问。

另外，官方也计划开源完整的训练代码，目前还在整理和工程化中，预计会在近期陆续放出，欢迎体验哈～
再次感谢在SenseNova-U1上做了很多工作，也欢迎后续继续交流反馈。

**Comment by Yidhar** [comment]
看了下技术报告 看来我的实现距离官方的版本有一定差距，感谢开源 我即刻修复

**Comment by Yidhar** [comment]
> 感谢这么详细的反馈和实验分享🙏🙏
> 
> 关于大家比较关心的训练细节，我们最近刚刚发布了[技术报告](https://github.com/OpenSenseNova/SenseNova-U1/blob/main/docs/pdf/SenseNOVA_U1.pdf)，里面对部分训练设计做一些说明，希望能解答部分疑问。
> 
> 另外，官方也计划开源完整的训练代码，目前还在整理和工程化中，预计会在近期陆续放出，欢迎体验哈～ 再次感谢在SenseNova-U1上做了很多工作，也欢迎后续继续交流反馈。

有一个问题 我参照技术报告使用了相同x-predict + v-loss 后 图像重建质量出现了明显的下滑 和格纹，远比x0_mse loss劣化严重是为什么
